### PR TITLE
When container had no layer data, cleanupContainer crashed

### PR DESCRIPTION
**Pull request by mYmNeo**
### Background

When I tried to test `TestCreateGrowRootfs` and `TestCreateShrinkRootfs` with the following command
`DOCKER_STORAGE_OPTS="dm.override_udev_sync_check=true,dm.basesize=10G,dm.loopdatasize=40G" DOCKER_GRAPHDRIVER="devicemapper" TESTFLAGS="-check.f DockerSuite.TestCreate*" make test-integration-cli`. I found the above two test were always skipped, but it should not be. Then I change some code to make the two test able to run, and the `TestCreateShrinkRootfs` crash the daemon with the following log

```
1062 time="2016-04-01T06:25:08.009764500Z" level=debug msg="devmapper: AddDevice(hash=7dc4c61b0cb29d5698ce16faca46feaa42d5937cfd5f850     e953938b18e231432-init basehash=7a5b71596cd0bc77f6c5b6c051f6005a14bce55cfe69ce59d780df1b5c1fc8e6)"
1063 time="2016-04-01T06:25:08.009795514Z" level=debug msg="devmapper: AddDevice(hash=7dc4c61b0cb29d5698ce16faca46feaa42d5937cfd5f850     e953938b18e231432-init basehash=7a5b71596cd0bc77f6c5b6c051f6005a14bce55cfe69ce59d780df1b5c1fc8e6) END"
1064 time="2016-04-01T06:25:08.009826278Z" level=error msg="Error saving dying container to disk: open /var/lib/docker/containers/d38     7c8c947c5cb2813421e6f9376cc50d3115058aba5cd27132ee000c29adb66/config.v2.json: no such file or directory"
1065 2016/04/01 06:25:08 http: panic serving @: runtime error: invalid memory address or nil pointer dereference
1066 goroutine 283 [running]:
1067 net/http.(*conn).serve.func1(0xc8216faa50, 0x7fdef83b9088, 0xc821854ce8)
1068   /usr/local/go/src/net/http/server.go:1287 +0xb5
1069 github.com/docker/docker/layer.(*layerStore).ReleaseRWLayer(0xc82093f700, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
1070   /go/src/github.com/docker/docker/layer/layer_store.go:520 +0xc3
1071 github.com/docker/docker/daemon.(*Daemon).cleanupContainer(0xc820503080, 0xc820018380, 0xc82032b901, 0x0, 0x0)
1072   /go/src/github.com/docker/docker/daemon/delete.go:126 +0x918
```

Because the `RWLayer` field is nil
### Solution

Before `ReleaseRWLayer` of a container, check validation of the `RWLayer` field to prevent daemon crash

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @aaronlehmann @tonistiigi ptal; also, is this something we want in the milestone, given that it can result in a panic?

**Comment by LK4D4** [comment]
@mYmNeo could you show full trace? I think your code is valid, but probably we have unnecessary code somewhere

**Comment by LK4D4** [comment]
LGTM
@mYmNeo but your third commit message doesn't explain what you did(fix), optionally you can add how you did it there. But for now I think it's ok.

**Comment by tonistiigi** [comment]
@thaJeztah First 2 are test fixes that we don't necessarily need for 1.11. The last one is needed if it is a panic but we need a better understanding on why this happened and probably a specific test for it then.

**Comment by thaJeztah** [comment]
@tonistiigi alright; so we would prefer the first two commits to be in their own PR?

**Comment by thaJeztah** [comment]
Just discussed with @tonistiigi, this is not a new issues, so we don't consider this a blocking issue for RC3, but may still include it in 1.11 GA if we have tests and more information about the cause

**Comment by tonistiigi** [comment]
@mYmNeo The panic likely comes from https://github.com/docker/docker/blob/master/daemon/create.go#L77 if there is an error on `setSecurityOptions` or `setRWLayer` (may also come from double delete). Like @LK4D4 said, it would be good to see the full trace to make sure of that. On the first look I'm not sure if we should check nil `RWLayer` or if this `defer` should just only happen after `setRWLayer()` has been called. cc @dmcgowan

**Comment by dmcgowan** [comment]
@tonistiigi I think I agree with you that cleanup should happen after the `setRWLayer`, we really shouldn't be doing operations on the `container` variable until after the RWLayer is set to prevent this. Since there is currently only one operation in between the cleanup and set RWLayer, it might make more sense to just handle anything from `setSecurityOptions` that needs cleanup if that returns an error..

**Comment by mYmNeo** [comment]
@LK4D4 @tonistiigi @dmcgowan This panic happened when `setRWLayer` failed. Because the command `docker create --storage-opt size=5G busybox` raised an error at https://github.com/docker/docker/blob/master/daemon/graphdriver/devmapper/deviceset.go#L1870

File: daemon/delete.go [hunk 1]
```diff
@@ -123,10 +123,12 @@ func (daemon *Daemon) cleanupContainer(container *container.Container, forceRemo
 		return fmt.Errorf("Unable to remove filesystem for %v: %v", container.ID, err)
 	}
 
-	metadata, err := daemon.layerStore.ReleaseRWLayer(container.RWLayer)
-	layer.LogReleaseMetadata(metadata)
-	if err != nil && err != layer.ErrMountDoesNotExist {
-		return fmt.Errorf("Driver %s failed to remove root filesystem %s: %s", daemon.GraphDriverName(), container.ID, err)
+	if container.RWLayer != nil {
```
Comment (tonistiigi) [inline daemon/delete.go]: Please add a comment that this is true when container creation fails and `RWLayer` has not been created yet.

**Comment by tonistiigi** [comment]
On second thought, I think the nil check is ok. LGTM after a comment is added.

File: daemon/delete.go (hunk 1 above)
Comment (mYmNeo) [inline daemon/delete.go]: Yeah, I added a comment
```
// When container creation fails and `RWLayer` has not been created yet, we
// do not call `ReleaseRWLayer`
```

**Comment by coolljt0725** [comment]
This also closes #21753

**Comment by shishir-a412ed** [comment]
This fixes only half of the problem. Agreed there is no need to `ReleaseRWLayer` if `RWLayer` is nil. This patch takes care of that. However when I do:
`docker run -it --storage-opt size=5G fedora /bin/bash` where my passed size < base device size.
The error from the device mapper (https://github.com/docker/docker/blob/master/daemon/graphdriver/devmapper/deviceset.go#L1876) will be thrown back to higher level layers.

Eventually it will reach here, https://github.com/docker/docker/blob/master/daemon/create.go#L90
When this will return the error. The defer func (`cleanupContainer`) will be executed. This will try to write the container config file (config.v2.json) indicating the dead container. However since the container.root does not exist.

Daemon will log this error message:
ERRO[0083] Error saving dying container to disk: open /var/lib/docker/containers/9843f4ff8443920c6e63e70fb28ac777b8a239d3369f7f10ef1b2b730a44ee56/config.v2.json: no such file or directory 

My Patch completes the solution.

File: integration-cli/requirements.go [hunk 2]
```diff
@@ -109,22 +114,50 @@ var (
 	}
 	NotOverlay = testRequirement{
 		func() bool {
-			cmd := exec.Command("grep", "^overlay / overlay", "/proc/mounts")
-			if err := cmd.Run(); err != nil {
-				return true
+			if daemonStorageDriver == "" {
+				// get daemon info
+				status, body, err := sockRequest("GET", "/info", nil)
```
Comment (cpuguy83) [inline integration-cli/requirements.go]: Can we extract this out into a helper fn?

Comment (mYmNeo) [inline integration-cli/requirements.go]: I moved the code into `docker_utils.go`, made it looks simple.

**Comment by icecrime** [comment]
@cpuguy83 @tonistiigi PTAL!

**Comment by tonistiigi** [comment]
LGTM

**Comment by cpuguy83** [comment]
LGTM

**Comment by tiborvass** [comment]
@mYmNeo would you mind rebasing please?

**Comment by tiborvass** [comment]
Carried by 21802

**Pull request closed without merging**